internal/urlutil/url.go

```diff
@@ -52,6 +57,55 @@ func ParseAndValidateURL(rawurl string) (*url.URL, error) {
 	return u, nil
 }
 
+type SharedURL struct {
+	*url.URL
+	initDone uint32
+	hostname func() string
```

Comment:
Would it be simpler to pre-compute the hostname?

```go
type SharedURL struct {
	*url.URL
	hostname string
}

func (s *SharedURL) Hostname() string {
	return s.hostname
}

var urlCache sync.Map // map[string]url.SharedURL
```

Does using singleflight with a sync.Map exhibit better performance than just a mutex and normal map? I've only ever used singleflight with IO-bound tasks.

Reply:
This can probably be improved, yeah